### DISCUSSION TITLE
Does Tycho support Target location type in a *.target?

### DISCUSSION THREAD

**Discussion by merks** [eclipse-tycho/tycho#973] [Q&A] [answered]
I was experimenting with PDE's new Target File support.  Something like this works in the IDE:
```
<location type="Target" uri="platform:/resource/target-platforms/m2e-target.target"/>
```
But I doubt that will work when Tycho sees this during a build.

The following works too, but only if Oomph is installed to provide/implement `file_uri`:
```
${file_uri:${project_loc:/target-platforms}}/m2e-target.target
```
Of course the above won't work in Tycho either.

A relative URI is not supported either.  And of course hard coding an absolute file: URI is unworkable.

So the only general thing that seems to be workable in the IDE to refer to another *.target in the workspace is a  platform:/resource URI, which I don't think user's will discover on their own.

So all this makes me wonder if this feature is supported in Tycho and if so what form of URI can it handle?

**Comment by laeubi** [comment, marked as answer]
> So all this makes me wonder if this feature is supported in Tycho and if so what form of URI can it handle?

The location type is supported, the usual idiom is to use:

` uri="file:${project_loc:/target-platforms}/<some file relative to the project 'target-platforms'>.target"`

see for example https://github.com/eclipse/tycho/tree/master/tycho-its/projects/target.references/target.refs

This is also what the PDE tooling suggests if you add such a target and open the drop-down for the target location (even though some more help like auto-complete for existing files would be useful).

Other common ways that work across IDE/Tycho are:

1. You can use https:// URLs for example https://raw.githubusercontent.com/eclipse/tycho/master/tycho-its/projects/target.usefile/jetty.target
2. You can use a maven deployed target e.g `mvn:groupid:artifactid:version:target`

> A relative URI is not supported either.

I have proposed these to PDE (and corresponding Tycho ticket) but it does not got much attention:

- https://bugs.eclipse.org/bugs/show_bug.cgi?id=578429
- #580 

> what form of URI can it handle

It can handle anything that resolves to a valid URI (tycho only supports a limited set of variables) that can be transformed into an URL and that URL can produce an InputStream to read from.

**Comment by merks** [comment]
That idiom doesn't work on Windows:

![image](https://user-images.githubusercontent.com/208716/169806919-fe0cb23e-1fc6-417a-917a-83c03d360750.png)

Yes, I noticed PDE suggested that but on Windows a file system path doesn't look like a URI, hence the need for `${file_uri:...}.` which comes form Oomph where there is often a need for a proper URL.

I did notice that an arbitrary URL works but that doesn't help me reference an target file elsewhere in the workspace (elsewhere in the clone), though it helps in that platform:/resource works as a result.

But in the end, there is no reasonable solution for referring to a *.target elsewhere in the workspace, nor even within the same project (relative URI), in a way that works in the IDE and also works with Tycho.  And because it doesn't work now, it's not likely to work any time soon.  That makes the feature, which is really **very** cool, unfortunately just not useful for my intended purpose...

Thanks for the quick answers!

**Reply by laeubi** [reply]
> That idiom doesn't work on Windows:

You need to prefix that file `file:` protocol! (maybe the suggestion is wrong?).

> the need for `${file_uri:...}.` which comes form Oomph

It probably would be a good idea to contribute this to platform, I think this is not much Oomph specific?

> I did notice that an arbitrary URL works but that doesn't help me reference an target file elsewhere in the workspace (elsewhere in the clone), though it helps in that platform:/resource works as a result.

The problem is that `platform:/resource` has not a real meaning inside a tycho build, but it *might* be possible to support this similar to how `project_loc` is supported (what makes some assumptions about possible matches).

> But in the end, there is no reasonable solution for referring to a *.target elsewhere in the workspace, nor even within the same project (relative URI), in a way that works in the IDE and also works with Tycho.

Please try the linked example it should work in IDE and in Tycho (if you are using latest PDE), at least if it do *not* work we need a bug for that and then will hopefully be fixed at least at the next release.

**Reply by laeubi** [reply]
That is what PDE offers me (Version: 2022-03 (4.23.0), Build id: 20220310-1457):
![grafik](https://user-images.githubusercontent.com/1331477/169810566-25cc665b-bb2e-4468-aae5-a8f9dd637b37.png)

**Reply by laeubi** [reply]
This is the [gerrit](https://git.eclipse.org/r/c/pde/eclipse.pde.ui/+/190949) that fixed this for windows and it [was verified](https://bugs.eclipse.org/bugs/show_bug.cgi?id=577184#c35) for 4.23 I20220223-1800.

**Comment by merks** [comment]
The dialog shows the error I showed above with the latest PDE, so I assumed this would not work, but later the location actually works, i.e., it resolved properly, so only a bug in that the dialog should be going through the logic provided by the Gerrit changes, but it does not.  But that's **very good news**!

May I ask now the referenced `${project_loc:...}` is resolved by Tycho.  Can it really be any other project in the same git clone (located anywhere in the folder tree)?

**Reply by laeubi** [reply]
> The dialog shows the error I showed above with the latest PDE, so I assumed this would not work, but later the location actually works, i.e., it resolved properly, so only a bug in that the dialog should be going through the logic provided by the Gerrit changes

That could be possible, and should be reported as a bug to PDE, the target reference support is rather new and the dialog could use some other improvements as well :-)

> May I ask now the referenced `${project_loc:...}` is resolved by Tycho. Can it really be any other project in the same git clone (located anywhere in the folder tree)?

It doesn't matter if it is in the same repository, it can be any *reactor* project, the reactor contains every project that is build in the current maven build and thus you can even combine projects from different git repos in the same reactor.

There are currently some limitations e.g.

- #538 

but if either the folder name or the (maven) project name or artifact id matches the project name it should be found, and tycho only supports the `${project_loc:<project name>}/some/other/file` syntax while eclipse also supports also full path as argument for `project_loc` but this should not be relevant here.

**Comment by merks** [comment]
**Very** cool!  This is super useful for my purposes!   As a big thank you, I've opened this issue and will contribute a PR shortly:

https://github.com/eclipse-pde/eclipse.pde/issues/121

**Comment by merks** [comment]
FYI, via https://bugs.eclipse.org/bugs/show_bug.cgi?id=580054 I extended Oomph's targlet infrastructure to support references to other *.target files and then in the generated *.target, consumable by Tycho, I'm making use of this capability to reference another *.target file.

**Reply by laeubi** [reply]
> I extended Oomph's targlet infrastructure to support references to other *.target files

Good to know even though I never used targlets (yet). :-)

> then in the generated *.target, consumable by Tycho, I'm making use of this capability to reference another *.target file

I know there is so much to do, but you could even add direct targlet support to tycho then it won't need to generate a separate target file :-)

**Comment by merks** [comment]
Even if you don't author targlets directly yourself,  you've set up development environments that use them, so you have used them.  :-P

I think supporting targlets directly in Tycho would be super challenging because part of the design point is to be able to specify source locators that induce a p2 repository from the contents of bundles and features in the file system, e.g., from the contents of a git clone, and to resolve the requirements of those without needing to duplicate those requirements somewhere else.   Furthermore, most users  create the targlet container using targlet tasks, so there is no *.target file in the first place.

So generating a Tycho-consumable *.target file is a good approach for capturing the information that is needed and already computed at development time...

**Comment by michirsch** [comment]
Thanks for the detailed explanation, @laeubi!

While looking into this, I also found the approach described in the official Tycho documentation, where you can list `.target` files directly in the POM using the `<target>` configuration block: https://tycho.eclipseprojects.io/doc/main/target-platform-configuration/target-platform-configuration-mojo.html#target

Now I’m wondering: what are the differences between this and the approach shown here using `<location type="Target">` inside a `.target` file?
Do both lead to the same result, or are there specific use cases or limitations that make one approach more suitable than the other?

Would really appreciate any clarification!

**Reply by laeubi** [reply]
@michirsch Usually you use the target-file because this works in the IDE and in the build, so you always are in sync.

Still there might be cases where one wants things managed in the pom.xml (e.g. because they are not important during development)... so there are always different options to archive the same goals. One example is that in PDE you can only ever activate *one* target at a time, while Tycho supports multiple files.

**Reply by michirsch** [reply]
@laeubi Thanks a lot for the quick and very helpful reply!

**Reply by michirsch** [reply]
Just wanted to add something in case anyone runs into the same situation:

When defining a list of `.target` artifacts in the POM (using multiple `<artifact>` entries), they need to be independent of each other. If, for example, `a.target` depends on `b.target`, this setup doesn’t seem to work in the POM-based approach.

In contrast, it does work if you define `b.target` as location inside `a.target`.
